// kintone-ui-component/lib/base/datetime/calendar/footer/style.js
export const CALENDAR_FOOTER_CSS = `
kuc-base-datetime-calendar-footer-1-17-0,
kuc-base-datetime-calendar-footer-1-17-0 *,
kuc-base-datetime-calendar-footer-1-17-0:lang(en),
kuc-base-datetime-calendar-footer-1-17-0:lang(en) * {
  font-family: sans-serif;
}
kuc-base-datetime-calendar-footer-1-17-0:lang(ja),
kuc-base-datetime-calendar-footer-1-17-0:lang(ja) * {
  font-family: "メイリオ", "Hiragino Kaku Gothic ProN", Meiryo,
    sans-serif;
}
kuc-base-datetime-calendar-footer-1-17-0:lang(zh),
kuc-base-datetime-calendar-footer-1-17-0:lang(zh) * {
  font-family: "微软雅黑", "Microsoft YaHei", "新宋体", NSimSun, STHeiti,
    Hei, "Heiti SC", sans-serif;
}
kuc-base-datetime-calendar-footer-1-17-0:lang(zh-TW),
kuc-base-datetime-calendar-footer-1-17-0:lang(zh-TW) * {
  font-family: "微軟正黑體","Microsoft JhengHei","新宋体",NSimSun,STHeiti,
    Hei,"Heiti SC",sans-serif;
}
.kuc-base-datetime-calendar-footer-1-17-0__group {
  display: flex;
  align-items: flex-end;
  box-sizing: border-box;
  padding: 0;
  height: 27px;
  white-space: nowrap;
  width: 272px;
}
.kuc-base-datetime-calendar-footer-1-17-0__group__button {
  background: transparent;
  border: 1px solid transparent;
  color: #3498db;
  cursor: pointer;
  font-size: 13px;
  outline: none;
}
.kuc-base-datetime-calendar-footer-1-17-0__group__button:hover {
  color: #217dbb;
}
.kuc-base-datetime-calendar-footer-1-17-0__group__button:focus {
  border: 1px solid #3498db;
  outline: none;
}
.kuc-base-datetime-calendar-footer-1-17-0__group__center {
  width: 100%;
}
`;
