// kintone-ui-component/lib/readonly-table/style.js
export const READ_ONLY_TABLE_CSS = `
  kuc-readonly-table-1-17-0 ,
  kuc-readonly-table-1-17-0  *,
  kuc-readonly-table-1-17-0:lang(en),
  kuc-readonly-table-1-17-0:lang(en) * {
    font-family: sans-serif;
  }
  kuc-readonly-table-1-17-0:lang(ja),
  kuc-readonly-table-1-17-0:lang(ja) * {
      font-family: "メイリオ", "Hiragino Kaku Gothic ProN", Meiryo,
      sans-serif;
  }
  kuc-readonly-table-1-17-0:lang(zh),
  kuc-readonly-table-1-17-0:lang(zh) * {
      font-family: "微软雅黑", "Microsoft YaHei", "新宋体", NSimSun, STHeiti,
      Hei, "Heiti SC", sans-serif;
  }
  kuc-readonly-table-1-17-0:lang(zh-TW),
  kuc-readonly-table-1-17-0:lang(zh-TW) * {
    font-family: "微軟正黑體","Microsoft JhengHei","新宋体",NSimSun,STHeiti,
      Hei,"Heiti SC",sans-serif;
  }
  kuc-readonly-table-1-17-0 {
    font-size: 14px;
    color: #333333;
    display: block;
  }
  kuc-readonly-table-1-17-0[hidden] {
    display: none;
  }
  .kuc-readonly-table-1-17-0__table {
    border-collapse: collapse;
    table-layout: fixed;
    width: 100%;
    display: inline-block;
  }
  .kuc-readonly-table-1-17-0__table__header {
    border-width: 0px 1px;
    border-color: var(--kuc-readonly-table-header-background-color, #3498db);
    border-style: solid;
  }
  .kuc-readonly-table-1-17-0__table__label {
    text-align: left;
    white-space: normal;
    overflow-wrap: anywhere;
    padding: 4px 0px;
  }
  .kuc-readonly-table-1-17-0__table__label[hidden] {
    display: none;
  }
  .kuc-readonly-table-1-17-0__table__label--no-column {
    overflow-wrap: break-word;
  }
  .kuc-readonly-table-1-17-0__table__header__cell {
    background-color: var(--kuc-readonly-table-header-background-color, #3498db);
    color: var(--kuc-readonly-table-header-color, #ffffff);
    height: var(--kuc-readonly-table-header-height, 40px);
    box-sizing: border-box;
    text-align: left;
    overflow: auto;
    white-space: nowrap;
    word-wrap: break-word;
    padding: 4px 8px;
    font-weight: 400;
    font-size: var(--kuc-readonly-table-header-font-size, 12px);
  }
  .kuc-readonly-table-1-17-0__table__header__cell[hidden] {
    display: none;
  }
  .kuc-readonly-table-1-17-0__table__body {
    vertical-align: top;
  }
  .kuc-readonly-table-1-17-0__table__body__row__cell-data {
    box-sizing: border-box;
    padding: 4px 8px;
    white-space: pre-wrap;
    overflow-wrap: break-word;
    border-color: #e3e7e8;
    border-style: solid;
    border-width: 1px;
  }
  .kuc-readonly-table-1-17-0__table__body__row__cell-data[hidden] {
    display: none;
  }
  .kuc-readonly-table-1-17-0__table__header__cell,
  .kuc-readonly-table-1-17-0__table__body__row__cell-data {
    scrollbar-width: none; /* Firefox */
  }
  .kuc-readonly-table-1-17-0__table__header__cell::-webkit-scrollbar,
  .kuc-readonly-table-1-17-0__table__body__row__cell-data::-webkit-scrollbar {
    width: 0; /* Safari and Chrome */
    display: none
  }
  .kuc-readonly-table-1-17-0__table__body__row__cell-data--html {
    white-space: normal;
  }
`;
