// kintone-ui-component/lib/radio-button/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html, svg } from "lit";
import { property, queryAll } from "lit/decorators.js";
import { ERROR_MESSAGE } from "../base/constant";
import { visiblePropConverter } from "../base/converter";
import { BaseError } from "../base/error";
import { createStyleOnHeader, dispatchCustomEvent, generateGUID, KucBase, } from "../base/kuc-base";
import { BaseLabel } from "../base/label";
import { validateArrayType, validateNumberType, validateProps, validateValueString, } from "../base/validator";
import { RADIOBUTTON_CSS } from "./style";
export { BaseError, BaseLabel };
let exportRadioButton;
(() => {
    exportRadioButton = window.customElements.get("kuc-radio-button-1-17-0");
    if (exportRadioButton) {
        return;
    }
    class KucRadioButton extends KucBase {
        constructor(props) {
            super();
            this.className = "";
            this.error = "";
            this.id = "";
            this.itemLayout = "horizontal";
            this.label = "";
            this.value = "";
            this.selectedIndex = -1;
            this.borderVisible = true;
            this.disabled = false;
            this.requiredIcon = false;
            this.visible = true;
            this.items = [];
            this._GUID = generateGUID();
            const validProps = validateProps(props);
            this._setInitialValue(validProps);
            Object.assign(this, validProps);
        }
        _setInitialValue(validProps) {
            const hasValue = "value" in validProps;
            const hasSelectedIndex = "selectedIndex" in validProps;
            if (!hasValue && hasSelectedIndex) {
                this.value = this._getValue(validProps) || "";
            }
        }
        shouldUpdate(changedProperties) {
            if (changedProperties.has("items")) {
                if (!validateArrayType(this.items)) {
                    this.throwErrorAfterUpdateComplete(ERROR_MESSAGE.ITEMS.IS_NOT_ARRAY);
                    return false;
                }
            }
            if (changedProperties.has("value")) {
                if (!validateValueString(this.value)) {
                    this.throwErrorAfterUpdateComplete(ERROR_MESSAGE.VALUE.IS_NOT_STRING);
                    return false;
                }
            }
            if (changedProperties.has("selectedIndex")) {
                if (!validateNumberType(this.selectedIndex)) {
                    this.throwErrorAfterUpdateComplete(ERROR_MESSAGE.SELECTED_INDEX.IS_NOT_NUMBER);
                    return false;
                }
            }
            return true;
        }
        _findItemToFocus() {
            let index = -1;
            for (let i = 0; i < this.items.length; i++) {
                const currentItem = this.items[i];
                if (currentItem.disabled)
                    continue;
                if (this.selectedIndex === i && currentItem.value === this.value) {
                    index = i;
                    continue;
                }
                if (index === -1) {
                    index = i;
                }
            }
            return index;
        }
        willUpdate(changedProperties) {
            if (changedProperties.has("value")) {
                if (this.value !== "")
                    return;
                this.selectedIndex = -1;
            }
        }
        _handleChangeInput(event) {
            event.stopPropagation();
            const inputEl = event.target;
            const value = inputEl.value;
            const index = inputEl.dataset.index || "0";
            const indexNumber = parseInt(index, 10);
            if (this.value === value && this.selectedIndex === indexNumber)
                return;
            const eventDetail = {
                oldValue: this.value,
                value: value,
            };
            this.value = value;
            this.selectedIndex = indexNumber;
            dispatchCustomEvent(this, "change", eventDetail);
        }
        _handleFocusInput(event) {
            const inputEl = event.target;
            const menuEl = inputEl.parentNode;
            menuEl.setAttribute("focused", "");
        }
        _handleBlurInput(event) {
            const inputEl = event.target;
            const menuEl = inputEl.parentNode;
            menuEl.removeAttribute("focused");
        }
        _getRadioIconSvgTemplate(disabled, checked) {
            return svg `
    <svg
      class="kuc-radio-button-1-17-0__group__select-menu__item__label__icon"
      width="21"
      height="21"
      viewBox="0 0 21 21"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="10.5"
        cy="10.5"
        r="10"
        fill="white"
        stroke="#e3e7e8" stroke-width="1"/>
      ${checked
                ? svg `<circle cx="10.5" cy="10.5" r="6.5" fill="${disabled ? "#e3e7e8" : "#3498db"}"/>`
                : ""}
    </svg>
  `;
        }
        _isCheckedItem(item, index) {
            if (!this.value)
                return this.selectedIndex === index;
            return item.value === this.value && this.selectedIndex === index;
        }
        _getItemTemplate(item, index) {
            const isCheckedItem = this._isCheckedItem(item, index);
            const isDisabledItem = item.disabled || this.disabled;
            const itemValue = item.value !== undefined ? item.value : "";
            const tabIndex = index === this._findItemToFocus() ? "0" : "-1";
            return html `
        <div
          class="kuc-radio-button-1-17-0__group__select-menu__item"
          itemLayout="${this.itemLayout}"
        >
          <input
            type="radio"
            aria-checked="${isCheckedItem ? "true" : "false"}"
            aria-describedby="${this._GUID}-error"
            data-index="${index}"
            id="${this._GUID}-item-${index}"
            class="kuc-radio-button-1-17-0__group__select-menu__item__input"
            name="${this._GUID}-group"
            value="${itemValue}"
            tabindex="${tabIndex}"
            aria-required="${this.requiredIcon}"
            ?disabled="${isDisabledItem}"
            @change="${this._handleChangeInput}"
            @focus="${this._handleFocusInput}"
            @blur="${this._handleBlurInput}"
          />
          <label
            class="kuc-radio-button-1-17-0__group__select-menu__item__label"
            for="${this._GUID}-item-${index}"
            >${this._getRadioIconSvgTemplate(isDisabledItem, isCheckedItem)}${item.label === undefined ? item.value : item.label}
          </label>
        </div>
      `;
        }
        update(changedProperties) {
            if (changedProperties.has("items") ||
                changedProperties.has("value") ||
                changedProperties.has("selectedIndex")) {
                this.selectedIndex = this._getSelectedIndex();
                this.value =
                    this._getValue({
                        items: this.items,
                        selectedIndex: this.selectedIndex,
                    }) || "";
            }
            super.update(changedProperties);
        }
        render() {
            return html `
        <div
          class="kuc-radio-button-1-17-0__group"
          role="radiogroup"
          aria-labelledby="${this._GUID}-group"
        >
          <div class="kuc-radio-button-1-17-0__group__label" ?hidden="${!this.label}">
            <kuc-base-label-1-17-0
              .text="${this.label}"
              .guid="${this._GUID}"
              .requiredIcon="${this.requiredIcon}"
            ></kuc-base-label-1-17-0>
          </div>
          <div
            class="kuc-radio-button-1-17-0__group__select-menu"
            ?borderVisible="${this.borderVisible}"
            itemLayout="${this.itemLayout}"
          >
            ${this.items.map((item, index) => this._getItemTemplate(item, index))}
          </div>
          <kuc-base-error-1-17-0
            .text="${this.error}"
            .guid="${this._GUID}"
            ?hidden="${!this.error}"
            ariaLive="assertive"
          ></kuc-base-error-1-17-0>
        </div>
      `;
        }
        updated() {
            this._inputEls.forEach((inputEl, idx) => {
                inputEl.checked =
                    this.value === inputEl.value && idx === this.selectedIndex;
            });
        }
        _getSelectedIndex() {
            if (!this.value) {
                if (this.items[this.selectedIndex])
                    return this.selectedIndex;
                return -1;
            }
            const firstIndex = this.items.findIndex((item) => item.value === this.value);
            if (firstIndex === -1)
                return -1;
            const selectedIndex = this.items.findIndex((item, index) => item.value === this.value && index === this.selectedIndex);
            return selectedIndex > -1 ? selectedIndex : firstIndex;
        }
        _getValue(validProps) {
            const _items = validProps.items || [];
            const _selectedIndex = validProps.selectedIndex === 0 || validProps.selectedIndex
                ? validProps.selectedIndex
                : -1;
            const item = _items[_selectedIndex];
            if (!item)
                return "";
            return item.value;
        }
    }
    __decorate([
        property({ type: String, reflect: true, attribute: "class" })
    ], KucRadioButton.prototype, "className", void 0);
    __decorate([
        property({ type: String })
    ], KucRadioButton.prototype, "error", void 0);
    __decorate([
        property({ type: String, reflect: true, attribute: "id" })
    ], KucRadioButton.prototype, "id", void 0);
    __decorate([
        property({ type: String })
    ], KucRadioButton.prototype, "itemLayout", void 0);
    __decorate([
        property({ type: String })
    ], KucRadioButton.prototype, "label", void 0);
    __decorate([
        property({ type: String })
    ], KucRadioButton.prototype, "value", void 0);
    __decorate([
        property({ type: Number })
    ], KucRadioButton.prototype, "selectedIndex", void 0);
    __decorate([
        property({ type: Boolean })
    ], KucRadioButton.prototype, "borderVisible", void 0);
    __decorate([
        property({ type: Boolean })
    ], KucRadioButton.prototype, "disabled", void 0);
    __decorate([
        property({ type: Boolean })
    ], KucRadioButton.prototype, "requiredIcon", void 0);
    __decorate([
        property({
            type: Boolean,
            attribute: "hidden",
            reflect: true,
            converter: visiblePropConverter,
        })
    ], KucRadioButton.prototype, "visible", void 0);
    __decorate([
        property({ type: Array })
    ], KucRadioButton.prototype, "items", void 0);
    __decorate([
        queryAll(".kuc-radio-button-1-17-0__group__select-menu__item__input")
    ], KucRadioButton.prototype, "_inputEls", void 0);
    window.customElements.define("kuc-radio-button-1-17-0", KucRadioButton);
    createStyleOnHeader(RADIOBUTTON_CSS);
    exportRadioButton = KucRadioButton;
})();
const RadioButton = exportRadioButton;
export { RadioButton };
