// kintone-ui-component/lib/time-picker/style.js
export const TIME_PICKER_CSS = `
kuc-time-picker-1-17-0,
kuc-time-picker-1-17-0 *,
kuc-time-picker-1-17-0:lang(en),
kuc-time-picker-1-17-0:lang(en) * {
  font-family: sans-serif;
}
kuc-time-picker-1-17-0:lang(ja),
kuc-time-picker-1-17-0:lang(ja) * {
  font-family: "メイリオ", "Hiragino Kaku Gothic ProN", Meiryo,
    sans-serif;
}
kuc-time-picker-1-17-0:lang(zh),
kuc-time-picker-1-17-0:lang(zh) * {
  font-family: "微软雅黑", "Microsoft YaHei", "新宋体", NSimSun, STHeiti,
    Hei, "Heiti SC", sans-serif;
}
kuc-time-picker-1-17-0:lang(zh-TW),
kuc-time-picker-1-17-0:lang(zh-TW) * {
  font-family: "微軟正黑體","Microsoft JhengHei","新宋体",NSimSun,STHeiti,
    Hei,"Heiti SC"
}
kuc-time-picker-1-17-0 {
  font-size: 14px;
  color: var(--kuc-time-picker-input-color, #333333);
  display: inline-table;
  max-width: var(--kuc-time-picker-input-width, 85px);
  width: var(--kuc-time-picker-input-width, 85px);
  vertical-align: top;
  line-height: 1.5;
}
.kuc-time-picker-1-17-0__group__input {
  position: relative;
}
kuc-time-picker-1-17-0[hidden] {
  display: none;
}
.kuc-time-picker-1-17-0__group {
  display: flex;
  flex-direction: column;
  border: none;
  padding: 0px;
  height: auto;
  margin: 0px;
}

.kuc-time-picker-1-17-0__group kuc-base-time-1-17-0 {
  display: inline-flex;
  flex-direction: column;
}

.kuc-time-picker-1-17-0__group .kuc-base-time-1-17-0__group {
  max-width: var(--kuc-time-picker-input-width, 85px);
  width: var(--kuc-time-picker-input-width, 85px);
  font-size: var(--kuc-time-picker-input-font-size, 14px);
  height: var(--kuc-time-picker-input-height, 40px);
  color: var(--kuc-time-picker-input-color, #333333);
}
.kuc-time-picker-1-17-0__group .kuc-base-time-1-17-0__group input[type=text].kuc-base-time-1-17-0__group__hours,
.kuc-time-picker-1-17-0__group .kuc-base-time-1-17-0__group input[type=text].kuc-base-time-1-17-0__group__minutes,
.kuc-time-picker-1-17-0__group .kuc-base-time-1-17-0__group input.kuc-base-time-1-17-0__group__suffix,
.kuc-time-picker-1-17-0__group .kuc-base-time-1-17-0__group--focus {
  color: var(--kuc-time-picker-input-color, #333333);
}
.kuc-time-picker-1-17-0__group__label {
  padding: 4px 0px 8px 0px;
  display: inline-block;
  white-space: nowrap;
}
.kuc-time-picker-1-17-0__group__label[hidden] {
  display: none;
}
`;
