// kintone-ui-component/lib/tooltip/style.js
export const TOOLTIP_CSS = `
  kuc-tooltip-1-17-0,
  kuc-tooltip-1-17-0 *,
  kuc-tooltip-1-17-0:lang(en),
  kuc-tooltip-1-17-0:lang(en) * {
    font-family: sans-serif;
  }
  kuc-tooltip-1-17-0:lang(ja),
  kuc-tooltip-1-17-0:lang(ja) * {
    font-family: "メイリオ", "Hiragino Kaku Gothic ProN", Meiryo,
      sans-serif;
  }
  kuc-tooltip-1-17-0:lang(zh),
  kuc-tooltip-1-17-0:lang(zh) * {
    font-family: "微软雅黑", "Microsoft YaHei", "新宋体", NSimSun, STHeiti,
      Hei, "Heiti SC", sans-serif;
  }
  kuc-tooltip-1-17-0:lang(zh-TW),
  kuc-tooltip-1-17-0:lang(zh-TW) * {
    font-family: "微軟正黑體","Microsoft JhengHei","新宋体",NSimSun,STHeiti,
      Hei,"Heiti SC",sans-serif;
  }
  .kuc-tooltip-1-17-0__group__title--hidden {
    display: none;
  }
  .kuc-tooltip-1-17-0__group {
    position: relative;
    display: inline-block;
  }
  .kuc-tooltip-1-17-0__group::after {
    position: absolute;
    right: -20%;
    top: 100%;
    left: -20%;
    display: block;
    height: calc(0.5em * 2);
  }
  .kuc-tooltip-1-17-0__group__title .kuc-tooltip-1-17-0__group__title__wrapper__text {
    max-width: var(--kuc-tooltip-width, 250px);
    width: var(--kuc-tooltip-width, auto);
    min-height: var(--kuc-tooltip-height, 32px);;
    height: var(--kuc-tooltip-height, auto);
    padding: 6px 8px;
    color: var(--kuc-tooltip-color, #ffffff);
    text-align: start;
    text-decoration: none;
    word-wrap: break-word;
    overflow: auto;
    white-space: normal;
    background-color: var(--kuc-tooltip-background-color, #000000);
    font-size: var(--kuc-tooltip-font-size);
    border-radius: 6px;
    box-shadow: 0 6px 16px 0 rgb(0 0 0 / 8%), 0 3px 6px -4px rgb(0 0 0 / 12%), 0 9px 28px 8px rgb(0 0 0 / 5%);
    box-sizing: border-box;
  }
  .kuc-tooltip-1-17-0__group__title {
    position: absolute;
    top: calc(100% + calc(0.5em * 2));
    left: 50%;
    transform: translateX(-50%);
    margin: 0;
    border-radius: 0.25em;
    color: var(--kuc-tooltip-color, #ffffff);
    width: max-content;
    z-index: 1000;
  }
  .kuc-tooltip-1-17-0__group.kuc-tooltip-1-17-0__group--top > .kuc-tooltip-1-17-0__group__title {
    top: unset;
    bottom: 100%;
  }
  .kuc-tooltip-1-17-0__group.kuc-tooltip-1-17-0__group--bottom > .kuc-tooltip-1-17-0__group__title {
    width: max-content;
    top: auto;
  }
  .kuc-tooltip-1-17-0__group.kuc-tooltip-1-17-0__group--left > .kuc-tooltip-1-17-0__group__title {
    width: max-content;
    height: fit-content;
    margin: auto 0;
    right: 100%;
    left: auto;
    top: 0;
    bottom: 0;
    transform: translateX(0);
  }
  .kuc-tooltip-1-17-0__group.kuc-tooltip-1-17-0__group--right > .kuc-tooltip-1-17-0__group__title {
    width: max-content;
    height: fit-content;
    margin: auto 0;
    left: 100%;
    top: 0;
    bottom: 0;
    transform: translateX(0);
  }
  .kuc-tooltip-1-17-0__group .kuc-tooltip-1-17-0__group__title__wrapper__arrow {
    border: 0.5em solid transparent;
    border-bottom-color: var(--kuc-tooltip-background-color, #000000);
  }
  .kuc-tooltip-1-17-0__group.kuc-tooltip-1-17-0__group--top > .kuc-tooltip-1-17-0__group__title .kuc-tooltip-1-17-0__group__title__wrapper__arrow {
    border-top-color: var(--kuc-tooltip-background-color, #000000);
    border-right-color: transparent;
    border-bottom-color: transparent;
    margin: auto 0;
  }
  .kuc-tooltip-1-17-0__group.kuc-tooltip-1-17-0__group--left > .kuc-tooltip-1-17-0__group__title .kuc-tooltip-1-17-0__group__title__wrapper__arrow {
    border-left-color: var(--kuc-tooltip-background-color, #000000);
    border-bottom-color: transparent;
  }
  .kuc-tooltip-1-17-0__group.kuc-tooltip-1-17-0__group--right > .kuc-tooltip-1-17-0__group__title .kuc-tooltip-1-17-0__group__title__wrapper__arrow {
    border-right-color: var(--kuc-tooltip-background-color, #000000);
    border-bottom-color: transparent;
    width: fit-content;
    height: fit-content;
    margin: auto 0;
    top: 0;
    bottom: 0;
    right: 100%;
    left: auto;
  }
  .kuc-tooltip-1-17-0__group .kuc-tooltip-1-17-0__group__title__wrapper {
    display: flex;
    align-items: center;
  }
  .kuc-tooltip-1-17-0__group.kuc-tooltip-1-17-0__group--top > .kuc-tooltip-1-17-0__group__title .kuc-tooltip-1-17-0__group__title__wrapper {
    flex-direction: column-reverse;
  }
  .kuc-tooltip-1-17-0__group.kuc-tooltip-1-17-0__group--bottom > .kuc-tooltip-1-17-0__group__title .kuc-tooltip-1-17-0__group__title__wrapper {
    flex-direction: column;
  }
  .kuc-tooltip-1-17-0__group.kuc-tooltip-1-17-0__group--left > .kuc-tooltip-1-17-0__group__title .kuc-tooltip-1-17-0__group__title__wrapper {
    flex-direction: row-reverse;
  }
`;
